### PR TITLE
Adding `--check` to run a serials network test against GitHub or GHES.

### PR DESCRIPTION
**Pull request by TingluoHuang**
Check github.com (Pass)
```
ting@htl-mac _layout % ./run.sh --check --url https://github.com/actions/runner --pat xxxxxxxxxxxxxx
**********************************************************************************************************************
**  Check:               Internet Connection
**  Description:         Make sure the actions runner have access to public internet.
**********************************************************************************************************************
**                                                                                                                  **
**                                            P A S S                                                               **
**                                                                                                                  **
**********************************************************************************************************************
** Log: /Users/ting/Desktop/runner/_layout/_diag/InternetCheck_20210113-200235-utc.log
**********************************************************************************************************************


**********************************************************************************************************************
**  Check:               GitHub Actions Connection
**  Description:         Make sure the actions runner have access to the GitHub Actions Service.
**********************************************************************************************************************
**                                                                                                                  **
**                                            P A S S                                                               **
**                                                                                                                  **
**********************************************************************************************************************
** Log: /Users/ting/Desktop/runner/_layout/_diag/ActionsCheck_20210113-200235-utc.log
**********************************************************************************************************************


**********************************************************************************************************************
**  Check:               Git Certificate/Proxy Validation
**  Description:         Make sure the git cli can access to GitHub.com or the GitHub Enterprise Server.
**********************************************************************************************************************
**                                                                                                                  **
**                                            P A S S                                                               **
**                                                                                                                  **
**********************************************************************************************************************
** Log: /Users/ting/Desktop/runner/_layout/_diag/GitCheck_20210113-200235-utc.log
**********************************************************************************************************************


**********************************************************************************************************************
**  Check:               Node.js Certificate/Proxy Validation
**  Description:         Make sure the node.js have access to GitHub.com or the GitHub Enterprise Server.
**********************************************************************************************************************
**                                                                                                                  **
**                                            P A S S                                                               **
**                                                                                                                  **
**********************************************************************************************************************
** Log: /Users/ting/Desktop/runner/_layout/_diag/NodeJsCheck_20210113-200235-utc.log
**********************************************************************************************************************
```

Check GHES: (Failed)
```
ting@htl-mac _layout % ./run.sh --check --url https://ting0113im1.ghesdev.com/actions/checkout --pat xxxxxxxxxx
**********************************************************************************************************************
**  Check:               Internet Connection
**  Description:         Make sure the actions runner have access to public internet.
**********************************************************************************************************************
**                                                                                                                  **
**                                            P A S S                                                               **
**                                                                                                                  **
**********************************************************************************************************************
** Log: /Users/ting/Desktop/runner/_layout/_diag/InternetCheck_20210113-202104-utc.log
**********************************************************************************************************************


**********************************************************************************************************************
**  Check:               GitHub Actions Connection
**  Description:         Make sure the actions runner have access to the GitHub Actions Service.
**********************************************************************************************************************
**                                                                                                                  **
**                                            F A I L                                                               **
**                                                                                                                  **
**********************************************************************************************************************
** Log: /Users/ting/Desktop/runner/_layout/_diag/ActionsCheck_20210113-202104-utc.log
** Help Doc: https://github.com/actions/runner/blob/main/docs/checks/actions.md
**********************************************************************************************************************


**********************************************************************************************************************
**  Check:               Git Certificate/Proxy Validation
**  Description:         Make sure the git cli can access to GitHub.com or the GitHub Enterprise Server.
**********************************************************************************************************************
**                                                                                                                  **
**                                            P A S S                                                               **
**                                                                                                                  **
**********************************************************************************************************************
** Log: /Users/ting/Desktop/runner/_layout/_diag/GitCheck_20210113-202104-utc.log
**********************************************************************************************************************


**********************************************************************************************************************
**  Check:               Node.js Certificate/Proxy Validation
**  Description:         Make sure the node.js have access to GitHub.com or the GitHub Enterprise Server.
**********************************************************************************************************************
**                                                                                                                  **
**                                            P A S S                                                               **
**                                                                                                                  **
**********************************************************************************************************************
** Log: /Users/ting/Desktop/runner/_layout/_diag/NodeJsCheck_20210113-202104-utc.log
**********************************************************************************************************************
```

### CODE REVIEW

File: src/Runner.Listener/Configuration/ConfigurationManager.cs [hunk 1]
```diff
@@ -509,13 +509,6 @@ private TaskAgent CreateNewAgent(string agentName, RSAParameters publicKey, ISet
             return agent;
         }
 
-        private bool IsHostedServer(UriBuilder gitHubUrl)
```
Comment (TingluoHuang) [inline src/Runner.Listener/Configuration/ConfigurationManager.cs:512]: moved to `UrlUtil.cs`

File: docs/checks/network.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+### Some lessons we learned in the past around things that can cause the runner to not working properly
+
+- Bug in the Dotnet framework that cause actions runner can't make Http request in a certain network environment.
+
+- Proxy/Firewall block certain HTTP method, like it block all POST and PUT calls which the runner will use to upload logs.
+
+- Proxy/Firewall only allows requests with certain user-agent to pass through and the actions runner user-agent is not in the allow list.
+
+- Proxy try to decrypt and exam HTTPS traffic for security purpose but cause the actions-runner fail to finish SSL handshake due to the leak of trusting proxy's CA.
```
Comment (ericsciple) [inline docs/checks/network.md:11]: lack

File: docs/checks/sslcert.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+3. Trust CA on different OS:
+    - Windows: https://docs.microsoft.com/en-us/skype-sdk/sdn/articles/installing-the-trusted-root-certificate
+    - macOS: ![trust ca cert](./../res/macOStrustCA.gif)
+    - Linux: Refer to the distribution documentation
+      1. RedHat: https://www.redhat.com/sysadmin/ca-certificates-cli
+      2. Ubuntu: http://manpages.ubuntu.com/manpages/focal/man8/update-ca-certificates.8.html
+      3. Google search: "trust ca certificate on [linux distribution]"
+      4. If all approaches failed, set environment variable `SSL_CERT_FILE` to the CA bundle `.pem` file we get. 
+    > To verity cert gets installed properly on Linux, you can try use `curl -v https://sitewithsslissue.com` and `pwsh -Command \"Invoke-WebRequest -Uri https://sitewithsslissue.com\"`
+### Trust CA certificate for Git CLI
```
Comment (ericsciple) [inline docs/checks/sslcert.md:48]: missing blank line before header

File: src/Misc/layoutbin/checkScripts/makeWebRequest.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
+const https = require('https')
+const http = require('http')
+const hostname = process.env['HOSTNAME']
+const port = process.env['PORT']
+const path = process.env['PATH']
+const pat = process.env['PAT']
+const proxyHost = process.env['PROXYHOST']
+const proxyPort = process.env['PROXYPORT']
+const proxyUsername = process.env['PROXYUSERNAME']
```
Comment (ericsciple) [inline src/Misc/layoutbin/checkScripts/makeWebRequest.js:9]: you might want to coalesce with `''` so we dont get `null` when string concat below

File: src/Runner.Listener/Checks/ActionsCheck.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            checkTasks.Add(HostContext.CheckHttpsRequests(actionsPipelinesServiceUrl, pat, expectedHeader: "x-vss-e2eid"));
+
+            var result = true;
+            while (checkTasks.Count > 0)
+            {
+                var finishedCheckTask = await Task.WhenAny<CheckResult>(checkTasks);
+                var finishedCheck = await finishedCheckTask;
+                result = result && finishedCheck.Pass;
+                await File.AppendAllLinesAsync(_logFile, finishedCheck.Logs);
+                checkTasks.Remove(finishedCheckTask);
```
Comment (ericsciple) [inline src/Runner.Listener/Checks/ActionsCheck.cs:83]: queue?

File: src/Runner.Listener/Checks/CheckUtil.cs [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+                    }
+                    else
+                    {
+                        env["PROXYUSERNAME"] = "";
+                        env["PROXYPASSWORD"] = "";
+                    }
+                }
+                else
+                {
+                    env["PROXYHOST"] = "";
```
Comment (ericsciple) [inline src/Runner.Listener/Checks/CheckUtil.cs:231]: do we need to worry about race condition between threads modifying env?

**Review by ericsciple — APPROVED** [review]

**Comment by ericsciple** [comment]
very cool

**Comment by jenschelkopf** [comment]
Hi @lucascosti, @TingluoHuang added a param to the runner that will help users diagnose issues with self-hosted runner connectivity to GitHub.com, GHES, or GHAE. Could you update the Actions docs to let people know about it?

**Comment by lucascosti** [comment]
> Could you update the Actions docs to let people know about it?

This is pretty cool! Thanks, @jenschelkopf! I've opened a docs release tracking issue to track adding docs: https://github.com/github/docs-content/issues/3765

**Pull request merged**